### PR TITLE
Unroll `map`, `mapi`, and `map2`, and `init`

### LINKED ISSUE #22: 🗞️ Unrolling?
I'm not sure if this is worth it, but I'll document it here anyways. The standard library is currently 2-unrolling a few functions, which means it fuses two recursive calls into one. The main reason is to match the performance, thanks to the careful benchmarking of @nojb. The relevant functions are:
- `init`
- `map`
- `mapi`
- `map2`

I haven't done the unrolling because the potential 20% slow-down on short lists hasn't become a bottleneck yet, but maybe we should do it when more people are using this library.

### PR DESCRIPTION
**Pull request by emmanueljs1**
Close https://github.com/RedPRL/ocaml-bwd/issues/22.

### CODE REVIEW

**Comment by emmanueljs1** [comment]
Tests pass with `dune test`.

File: src/BwdNoLabels.ml [hunk 1]
```diff
@@ -57,15 +57,15 @@ let nth_opt xs i =
...
-      if i >= len then Emp
-      else if i = len - 1 then Snoc (Emp, f i)
+    let[@tail_mod_cons] rec go i last =
+      if i > last then Emp
+      else if i = last then Snoc (Emp, f i)
       else
         let v1 = f i in
         let v2 = f (i + 1) in
-        Snoc (Snoc ((go[@tailcall]) (i+2), v2), v1)
+        Snoc (Snoc ((go[@tailcall])(i+2)(last), v2), v1)
```
Comment (favonia) [inline src/BwdNoLabels.ml]:
```suggestion
        Snoc (Snoc ((go[@tailcall]) (i + 2) last, v2), v1)
```
for style consistency

**Comment by favonia** [comment]
@emmanueljs1 Thank you for the PR. Could you possibly reformat the code with ocp-indent? Then I think we are good to go!

**Comment by emmanueljs1** [comment]
> @emmanueljs1 Thank you for the PR. Could you possibly reformat the code with ocp-indent? Then I think we are good to go!

Done!

**Review by favonia — APPROVED** [review]
Thanks!

**Pull request merged**